sql/core/src/main/scala/org/apache/spark/sql/execution/command/AlterTableCommandParser.scala

```diff
@@ -0,0 +1,410 @@
+                            vals.map(n => cleanAndUnquoteString(n.text))
+                        }
+                    }
+                }
+
+                val storedAsDirs = stored match {
+                  case Token("TOK_STOREDASDIRS", Nil) :: Nil => true
+                  case _ => false
+                }
+
+                (cols, values, storedAsDirs)
+            }
+        }
+        val (cols, values, storedAsDirs) = skewedArgs
+        AlterTableSkewed(
+          tableIdent,
+          cols,
+          values,
+          storedAsDirs,
+          notSkewed = false,
+          notStoredAsDirs = false)(node.source)
+
+      case Token("TOK_ALTERTABLE_SKEWED_LOCATION",
+      Token("TOK_SKEWED_LOCATIONS",
+      Token("TOK_SKEWED_LOCATION_LIST", locationMaps) :: Nil) :: Nil) :: _ =>
+        val skewedMaps = locationMaps.map {
+          case Token("TOK_SKEWED_LOCATION_MAP", key :: value :: Nil) =>
+            val k = key match {
+              case Token(const, Nil) => Seq(cleanAndUnquoteString(const))
+              case Token("TOK_TABCOLVALUES", values :: Nil) =>
+                values match {
+                  case Token("TOK_TABCOLVALUE", vals) =>
+                    vals.map(n => cleanAndUnquoteString(n.text))
+                }
+            }
+            (k, cleanAndUnquoteString(value.text))
+        }.toMap
+        AlterTableSkewedLocation(tableIdent, skewedMaps)(node.source)
+
+      case Token("TOK_ALTERTABLE_ADDPARTS", addPartsArgs) :: _ =>
+        val (allowExisting, parts) = addPartsArgs match {
+          case Token("TOK_IFNOTEXISTS", Nil) :: others => (true, others)
+          case _ => (false, addPartsArgs)
+        }
+        val partitions: ArrayBuffer[(Map[String, Option[String]], Option[String])] =
+          new ArrayBuffer()
+        var currentPart: Map[String, Option[String]] = null
+        parts.map {
+          case t @ Token("TOK_PARTSPEC", partArgs) =>
+            if (currentPart != null) {
+              partitions += ((currentPart, None))
+            }
+            currentPart = parsePartitionSpec(t).get
+          case Token("TOK_PARTITIONLOCATION", loc :: Nil) =>
+            val location = unquoteString(loc.text)
+            if (currentPart != null) {
+              partitions += ((currentPart, Some(location)))
+              currentPart = null
+            } else {
+              // We should not reach here
+              throw new AnalysisException("Partition location must follow a partition spec.")
+            }
+        }
+        if (currentPart != null) {
+          partitions += ((currentPart, None))
+        }
+        AlterTableAddPartition(tableIdent, partitions, allowExisting)(node.source)
+
+      case Token("TOK_ALTERTABLE_RENAMEPART", partArg :: Nil) :: _ =>
+        val Some(newPartition) = parsePartitionSpec(partArg)
+        AlterTableRenamePartition(tableIdent, partition.get, newPartition)(node.source)
+
+      case Token("TOK_ALTERTABLE_EXCHANGEPARTITION",
```

Comment:
We assume that `TOK_ALTERTABLE_EXCHANGEPARTITION` is followed by `TOK_PARTSPEC` and `TOK_TABNAME`. Why not let `parsePartitionSpec` and `extractTableIdent` validate this for us? This would save us some trouble in the match.